refactor(linter/plugins): introduce methods for deep cloning JSON values

overlookmotel opened the pull request:
Add methods for deep cloning JSON values, objects, and arrays. These are used in the next PR (#16930).

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16962?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#16930** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16930?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16962** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16962?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16962?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#16961** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16961?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16960** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16960?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16959** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16959?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16958** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16958?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=16962).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

nrayburn-tech commented:
No idea of perf, but was https://developer.mozilla.org/en-US/docs/Web/API/Window/structuredClone considered?

overlookmotel committed:
refactor(linter/plugins): introduce methods for deep cloning JSON values (#16962)

Add methods for deep cloning JSON values, objects, and arrays. These are used in the next PR (#16930).
